src/re/multi_grid/correlated_field.py

```diff
@@ -0,0 +1,119 @@
+
+        Parameters
+        ----------
+        grid: Grid
+            Storage object wiht instructions on how pixels are spaces.
+        kernel: dict
+            Parameters for the kernel. Currently supports `kind='matern'` with the
+            following additional arguments
+
+            - scale: tuple or callable or float -- Prior scale of the kernel.
+            - cutoff: tuple or callable or float -- Prior cutoff of the covariance kernel.
+            - loglogslope: tuple or callable or float -- Prior power-law slope of the covariance kernel correlating modes.
+            - n_integrate = 2_000 -- Number of integration points for the harmonic transformation.
+            - n_interpolate = 128 -- Number of interpolation points for interpolating the kernel.
+            - interpolation_dists_min_max = (1e-5, 1e2) -- Interpolation range.
+
+            alternatively, `kind='fixed'` with
+
+            - covariance: callable -- Covariance kernel function.
+
+            is supported.
+        offset: tuple or allable or float
+            Prior shift from zero in addition to the field intrinsic random shift.
+        window_size:
+            Shape of the refinement matrices. Its value is automatically inferred
+            from the padding in the grid or set to sane defaults if left unspecified.
+        compress:
+            Arguments for compressing the refinement matrices. If this is set to an
+            empty dictionary, no compression is applied.
+        prexi: str
+            Prefix for the latent parameter names.
+        """
+        from .matern import MaternCovariance
+
+        self.grid = grid
+        shapes = [
+            ShapeWithDtype(self.grid.at(lvl).shape, jnp.float_)
+            for lvl in range(grid.depth + 1)
+        ]
+        self._name_exc = str(prefix) + "excitations"
+        domain = {self._name_exc: shapes}
+
+        DEFAULT_KERNEL_KIND = "matern"
+        if not isinstance(kernel, dict):
+            raise TypeError(f"invalid kernel type; got {kernel!r}")
+        fixed_kernel = False
+        kernel = kernel.copy()
+        kernel_kind = kernel.pop("kind", DEFAULT_KERNEL_KIND).lower()
+        if kernel_kind == "matern":
+            covariance = MaternCovariance(**kernel, prefix=prefix)
+        elif kernel_kind == "fixed":
+            fixed_kernel = True
+            covariance = kernel["covariance"]
+        else:
+            raise ValueError(f"kernel {kernel_kind!r} not supported")
+        self.covariance = covariance
+        self.fixed_kernel = fixed_kernel
+        if not self.fixed_kernel:
+            domain |= self.covariance.domain
+
+        if isinstance(offset, (tuple, list)):
+            offset = NormalPrior(*offset, name=prefix + "offset")
+        elif callable(offset) or not isinstance(offset, float):
+            raise ValueError
+        self.offset = offset
+        if callable(self.offset):
```

Comment:
Check for is instance of `jft.Model` instead? Ensure that domain exists. I don't think we need to support arbitrary callables in place of models here, right?

Reply:
Good point. `offset` should probably always be a model or a float. For consistency sake it does make sense to allow for arbitrary callables as to be consistent with the non-parameteric harmonic CFM. I've made the parsing more flexible to check for models and callables.